compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp

```diff
@@ -903,10 +905,34 @@ std::tuple<int64_t, int64_t, int64_t> DataTiledMMAAttr::getMNKShape() const {
           opaqueLayout.kSize * getUnrollK()};
 }
 
+static SmallVector<int64_t>
+filterSwizzledShape(const TileSwizzle &swizzle,
+                    std::function<bool(TileSwizzle::Dim)> predicate) {
+  SmallVector<int64_t> shape;
+  for (auto e : swizzle.expandShape) {
+    for (auto d : e) {
+      shape.push_back(predicate(d) ? d.size : 1);
+    }
+  }
+  applyPermutationToVector(shape, swizzle.permutation);
+  return shape;
+}
+
 std::tuple<VectorType, VectorType, VectorType>
 DataTiledMMAAttr::getABCVectorTypes() const {
-  return MMAAttr::get(getContext(), getIntrinsic().getValue())
-      .getABCVectorTypes();
+  auto [A, B, C] = getABCElementTypes();
+  SmallVector<Type> elementTypes = {A, B, C};
+  auto getVectorType = [=](MMAFragment fragment) -> VectorType {
+    auto swizzle = getSwizzle(*this, fragment);
+    SmallVector<int64_t> shape =
+        filterSwizzledShape(swizzle, [](TileSwizzle::Dim d) {
+          return d.kind != TileSwizzle::Dim::Kind::CrossThread;
+        });
+    Type elementType = elementTypes[static_cast<int>(fragment)];
+    return VectorType::get(shape, elementType);
+  };
+  return {getVectorType(MMAFragment::Lhs), getVectorType(MMAFragment::Rhs),
+          getVectorType(MMAFragment::Acc)};
```

Comment:
Maybe explicitly pass in `A`, `B`, and `C` element types? Then you don't need to static_cast the fragment to an index, and it only costs a few extra chars here to add the parameter.